Add draft support for CLI in Server extension

ChrisChinchilla opened the pull request:
OK, here we go… An attempt to merge the CLI and Server extensions together. this is very much draft and missing a few things, but wanted to get input o if this is the right direction at least.

ChrisChinchilla commented on src/features/vsProvider.ts:
```diff
@@ -152,21 +218,48 @@ export default class ValeServerProvider implements vscode.CodeActionProvider {
+    const binaryLocation = readBinaryLocation();
+    const configLocation = readFileLocation()!;
+    // TODO: Hmm
+    this.stylesPath = "/Users/chrisward/Workspace/TesttheDocs";
```
This in particular needs thought. I am not sure how to handle the equivalent in the CLI version, so it's hard coded right now while I was testing. To make this work properly, the path needs to follow the usual Vale CLi scoping rules.

jdkato replied:
You can get the path by inspecting the output of `ls-config` once you have the `binaryLocation` and `configLocation`:

```console
vale --config="/Users/jdkato/Desktop/docs/.vale.ini" ls-config
```

jdkato commented:
This looks good overall. 

By the time we merge, though, I would like `vale-server.useCLI` to be `false` by default. I think it's more reasonable to ask CLI users to change a setting than the (paying) server users.

ChrisChinchilla commented:
@jdkato seems reasonable, I think it was just easier for me to test this way. I'll do some more work soon, and then we need to look at migration. I also want to do some testing.

> This looks good overall.
> 
> By the time we merge, though, I would like `vale-server.useCLI` to be `false` by default. I think it's more reasonable to ask CLI users to change a setting than the (paying) server users.

ChrisChinchilla commented:
@jdkato Addressed the changes! Needs a bit of a tidy up now, and I want to update the read me too. Then I also need to transfer my issues and we can handle the marketplace steps.

jdkato commented:
Sounds good 👍

ChrisChinchilla commented:
@jdkato I have an updated version, but I am currently on a hotel wifi for the next few days that seems to be blocking port 22 which is problematic… Will push when I can. As part of the coming changes I changed the name of the extension. Open for discussion on this, I changed it to `vale-vscode` as that makes the most sense, but it is the name of the extension I maintain I realise (totally fine with it ending up as `errata-ai/vale-vscode`. However, we will have to make one set of extension users move anyway, so in either direction the name change happens, we need to get people to update things.

ChrisChinchilla commented:
@jdkato And updates to code pushed

jdkato commented:
I would *really* like to avoid changing the extension's name, as that is effectively creating an entirely new extension from a marketplace standpoint.

We *could* simply change the `displayName` to something like "Vale and Vale Server", but I think we can use the `description` to clarify the purpose of the extension&mdash;e.g., something like

> The official Visual Studio Code extension for Vale and Vale Server.

Ultimately, I'd like to view this PR as adding support for the Vale CLI to the existing extension for Vale Server; not creating a new extension that supports both. I know this is somewhat of a semantic difference but, since *all* existing users of this extension are already using Vale Server, I'd like to avoid introducing breaking or confusing changes from *that* perspective.

This would also mean that you don't have to explicitly change anything about your current extension either (besides, perhaps, a notice about the changes here). This way, we don't break either extension and we don't directly force either set of users to "move" to a new extension.

jdkato commented:
I'd like to get this merged so I can start work on a new release without introducing conflicts here, could you just reset to commit 6c7613370b3359a1e306f098648e875875e4e2fd and I'll merge?

Thanks!

ChrisChinchilla commented:
I was under the impression we were creating one extension that does both, not adding support to the server extension. With the intention of bringing less confusion to users.

The CLI extension has more users, and I've spoken to a lot of people who use it. If that is the extension we retire (which was the impression I had), we also need to migrate those users. I'm not sure if a miss-leading name will help ease that transition and confusion, so i'd like to get this right.

I agree that changing the name, while an ideal thing to do, won't help that confusion, but let me make a pass at making things as clear as possible in the display name and description.

Another option is to actually use the CLI extension name, move this code there, and transfer ownership of it. That means the smaller group of people have to migrate. But i'd really like to only have one extension that does everything and helps people the most, and we should focus on that first and foremost.

jdkato commented:
> I was under the impression we were creating one extension that does both, not adding support to the server extension. With the intention of bringing less confusion to users.

Yes, I think our goals are very similar, if not the same. I'm simply of the opinion that we should, if at all possible, avoid breaking an existing extension (for either side of users).

---

In any case, the changes here look good enough for now -- I'll ping you before I cut the release, though, to make sure we're still on the same page.

ChrisChinchilla commented:
OK @jdkato I'll handle getting users off the other extension, there's no rush there. I'll also transfer some issues to this repo for improvements I want to add.
